arcgis-ios-sdk-samples/Utility network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift

```diff
@@ -180,75 +268,78 @@ class PerformValveIsolationTraceViewController: UIViewController {
-        UIApplication.shared.showProgressHUD(message: "Running isolation trace…")
-        utilityNetwork.trace(with: parameters) { [weak self] (traceResults, error) in
+    func trace(completion: @escaping () -> Void) {
+        guard let configuration = makeTraceConfiguration(category: selectedCategory) else {
+            setStatus(message: "Failed to get trace configuration.")
+            return
+        }
+        traceParameters.traceConfiguration = configuration
+        
+        utilityNetwork.trace(with: traceParameters) { [weak self] traceResults, error in
             guard let self = self else { return }
-            let categoryName = self.selectedCategory!.name.lowercased()
             if let elementTraceResult = traceResults?.first as? AGSUtilityElementTraceResult,
                !elementTraceResult.elements.isEmpty {
-                UIApplication.shared.showProgressHUD(message: "Trace completed. Selecting features…")
-                self.selectFeatures(in: elementTraceResult.elements) { [weak self] in
-                    self?.setStatus(message: "Trace with \(categoryName) category completed.")
-                    UIApplication.shared.hideProgressHUD()
+                self.selectFeatures(in: elementTraceResult.elements) {
+                    if let categoryName = self.selectedCategory?.name.lowercased() {
+                        self.setStatus(message: "Trace with \(categoryName) category completed.")
+                    } else {
+                        self.setStatus(message: "Trace with filter barriers completed.")
+                    }
                 }
             } else if let error = error {
-                self.setStatus(message: "Trace with \(categoryName) category failed.")
+                self.setStatus(message: "Trace failed.")
                 self.presentAlert(error: error)
-                UIApplication.shared.hideProgressHUD()
             } else {
-                self.setStatus(message: "Trace with \(categoryName) category completed with no output.")
+                self.setStatus(message: "Trace completed with no output.")
+            }
+            completion()
+        }
+    }
+    
+    @IBAction func traceResetButtonTapped(_ button: UIBarButtonItem) {
+        switch traceCompleted {
+        case true:
+            // Reset the trace if it is already completed
+            clearLayersSelection()
+            traceParameters.filterBarriers.removeAll()
+            parametersOverlay.graphics.removeAllObjects()
+            traceCompleted = false
+            selectedCategory = nil
+            // Add back the starting location.
+            addGraphic(for: startingLocationPoint, traceLocationType: "starting point")
+            mapView.setViewpointCenter(startingLocationPoint, scale: 3000, completion: nil)
+            // Set UI state.
+            setStatus(message: "Tap on the map to add filter barriers, or run the trace directly without filter barriers.")
```

Comment:
```suggestion
            setStatus(message: "Tap on the map to add filter barriers or run the trace directly without filter barriers.")
```
In the README you mentioned that the filter barriers are configured when a category is selected. I think that should be expressed here as well.

Reply:
I hope today's demo helped to clear stuff up! Please let me know if further edit is needed in these status messages. 🙂 

> Looks good overall! What're your thoughts about adding an undo button in case the user wants to remove a filter barrier or go back to the category selection workflow? The toolbar seems full already but the only other way to exit that workflow is to trace and reset or restart the sample.

I was thinking about adding a reset button, but don't want to overcrowd the toolbar. I tend to keep the current UI TBH 😆